File: infrastructures/ddd/csharp/src/Infrastructure/Repositories/Impl/ExampleQueryRepository.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using OddlyDdd.Infrastructure.Persistence.Read;

namespace OddlyDdd.Infrastructure.Repositories.Impl
{
    /// <summary>
    /// Example query repository implementation demonstrating CQRS query side patterns.
    /// Query repository implementations:
    /// - Live in /Infrastructure/Repositories/Impl/
    /// - Return ReadEntities directly (NO BMO mapping)
    /// - Optimized for query performance
    /// - May use denormalized views
    /// 
    /// Note: This is a template. Actual implementation would use a real database
    /// (Entity Framework, MongoDB.Driver, Dapper, etc.)
    /// </summary>
    public class ExampleQueryRepository : IExampleQueryRepository
    {
        // In real implementation: private readonly DbContext _dbContext;
        // Or: private readonly IMongoCollection<ExampleReadEntity> _collection;

        public ExampleQueryRepository()
        {
        }

        /// <summary>
        /// Finds an example by ID.
        /// Returns ReadEntity directly - no BMO mapping.
        /// </summary>
        public async Task<ExampleReadEntity?> FindByIdAsync(string p_id)
        {
            // Query database
            // Example with EF Core: return await _dbContext.ExamplesView.FindAsync(p_id);
            // Example with MongoDB: return await _collection.Find(x => x.Id == p_id).FirstOrDefaultAsync();
            
            // Simulate async operation
            await Task.CompletedTask;
            return null;
        }

        /// <summary>
        /// Lists all examples.
        /// Returns ReadEntities directly.
        /// </summary>
        public async Task<IReadOnlyList<ExampleReadEntity>> ListAllAsync()
        {
            // Query database
            // Example with EF Core: return await _dbContext.ExamplesView.ToListAsync();
            // Example with MongoDB: return await _collection.Find(_ => true).ToListAsync();
            
            // Simulate async operation
            await Task.CompletedTask;
            return new List<ExampleReadEntity>();
        }

        /// <summary>
        /// Lists examples with pagination.
        /// Returns ReadEntities directly.
        /// </summary>
        public async Task<IReadOnlyList<ExampleReadEntity>> ListAsync(int p_skip, int p_take)
        {
            // Query database with pagination
            // Example with EF Core: return await _dbContext.ExamplesView.Skip(p_skip).Take(p_take).ToListAsync();
            // Example with MongoDB: return await _collection.Find(_ => true).Skip(p_skip).Limit(p_take).ToListAsync();
            
            // Simulate async operation
            await Task.CompletedTask;
            return new List<ExampleReadEntity>();
        }

        /// <summary>
        /// Counts all examples.
        /// </summary>
        public async Task<long> CountAsync()
        {
            // Count in database
            // Example with EF Core: return await _dbContext.ExamplesView.CountAsync();
            // Example with MongoDB: return await _collection.CountDocumentsAsync(_ => true);
            
            // Simulate async operation
            await Task.CompletedTask;
            return 0;
        }

        /// <summary>
        /// Finds examples by owner ID.
        /// Custom query method.
        /// </summary>
        public async Task<IReadOnlyList<ExampleReadEntity>> FindByOwnerIdAsync(string p_ownerId)
        {
            // Query database by owner
            // Example with EF Core: return await _dbContext.ExamplesView.Where(x => x.OwnerId == p_ownerId).ToListAsync();
            // Example with MongoDB: return await _collection.Find(x => x.OwnerId == p_ownerId).ToListAsync();
            
            // Simulate async operation
            await Task.CompletedTask;
            return new List<ExampleReadEntity>();
        }

        /// <summary>
        /// Finds active examples only.
        /// Custom query method demonstrating filtering.
        /// </summary>
        public async Task<IReadOnlyList<ExampleReadEntity>> FindActiveExamplesAsync(int p_skip, int p_take)
        {
            // Query database for active examples
            // Example with EF Core: return await _dbContext.ExamplesView.Where(x => x.IsActive).Skip(p_skip).Take(p_take).ToListAsync();
            // Example with MongoDB: return await _collection.Find(x => x.IsActive).Skip(p_skip).Limit(p_take).ToListAsync();
            
            // Simulate async operation
            await Task.CompletedTask;
            return new List<ExampleReadEntity>();
        }
    }
}

File: infrastructures/ddd/typescript/src/infrastructure/repositories/impl/ExampleQueryRepository.ts
import { IExampleQueryRepository } from '../IExampleQueryRepository';
import { ExampleReadEntity } from '../../persistence/read/ExampleReadEntity';

/**
 * Example query repository implementation demonstrating read operations.
 * Repository implementations:
 * - Live in /infrastructure/repositories/impl/
 * - Implement repository interface from parent directory
 * - Work with ReadEntities
 * - Return ReadEntities directly (NO BMO mapping for performance)
 * - Optimized for query operations
 * 
 * This is a demonstration implementation.
 * In a real application, you would inject your database context/client
 * and query read-optimized collections/views.
 */
export class ExampleQueryRepository implements IExampleQueryRepository {
  // In a real implementation, inject database context here:
  // constructor(private readonly _dbContext: DatabaseContext)

  constructor() {
    // Initialization
  }

  /**
   * Finds an example by its identifier.
   * Returns ReadEntity directly (no BMO mapping).
   */
  public async FindByIdAsync(p_id: string): Promise<ExampleReadEntity | null> {
    // In a real implementation, query read collection/view:
    // const entity = await this._dbContext.ExamplesView.findOne({ Id: p_id });
    // return entity;
    
    console.log('[ExampleQueryRepository] FindByIdAsync - ID:', p_id);
    return null; // Demonstration only
  }

  /**
   * Lists examples with pagination.
   * Returns ReadEntities directly (no BMO mapping).
   */
  public async ListAsync(p_skip: number, p_take: number): Promise<ExampleReadEntity[]> {
    // In a real implementation, query read collection/view:
    // const entities = await this._dbContext.ExamplesView
    //   .find()
    //   .skip(p_skip)
    //   .limit(p_take)
    //   .toArray();
    // return entities;
    
    console.log('[ExampleQueryRepository] ListAsync - Skip:', p_skip, 'Take:', p_take);
    return []; // Demonstration only
  }

  /**
   * Counts total number of examples.
   */
  public async CountAsync(): Promise<number> {
    // In a real implementation, count from read collection/view:
    // const count = await this._dbContext.ExamplesView.countDocuments();
    // return count;
    
    console.log('[ExampleQueryRepository] CountAsync');
    return 0; // Demonstration only
  }

  /**
   * Lists active examples with pagination.
   */
  public async ListActiveAsync(p_skip: number, p_take: number): Promise<ExampleReadEntity[]> {
    // In a real implementation, query read collection/view:
    // const entities = await this._dbContext.ExamplesView
    //   .find({ IsActive: true })
    //   .skip(p_skip)
    //   .limit(p_take)
    //   .toArray();
    // return entities;
    
    console.log('[ExampleQueryRepository] ListActiveAsync - Skip:', p_skip, 'Take:', p_take);
    return []; // Demonstration only
  }

  /**
   * Lists examples by owner with pagination.
   */
  public async ListByOwnerAsync(
    p_ownerId: string,
    _p_skip: number,
    _p_take: number
  ): Promise<ExampleReadEntity[]> {
    // In a real implementation, query read collection/view:
    // const entities = await this._dbContext.ExamplesView
    //   .find({ OwnerId: p_ownerId })
    //   .skip(p_skip)
    //   .limit(p_take)
    //   .toArray();
    // return entities;
    
    console.log('[ExampleQueryRepository] ListByOwnerAsync - OwnerId:', p_ownerId);
    return []; // Demonstration only
  }
}
